(PUP-5768) Fix problematic usage of ENV for Windows

glennsarti opened the pull request:
The ENV class in Ruby is known to corrupt environment variables on Windows which contain UTF8 characters in their names and values.

This PR adds an operating system safe way of accessing environment variables as part of the Puppet::Util module, modifies the places in Puppet where it could potentially corrupt environment variables and adds tests to guard against the behaviour in the future.

Iristyle commented on lib/puppet/test/test_helper.rb:
```diff
@@ -121,7 +121,13 @@ def self.before_each_test()
 
       # The process environment is a shared, persistent resource.
-      $old_env = ENV.to_hash
+      # Can not use Puppet.features.microsoft_windows? as it may be mocked out in a test.  This can cause test recurring test failures
```
typo

Iristyle commented on spec/unit/util_spec.rb:
```diff
@@ -487,7 +488,7 @@ def get_mode(file)
         it "should fall back if no extension matches" do
-          ENV.stubs(:[]).with('PATH').returns(base)
+          Puppet::Util.stubs(:env).with('PATH').returns(base)
           ENV.stubs(:[]).with('PATHEXT').returns(".EXE")
```
Did you miss a stub change here?

glennsarti replied:
No.  The code uses ENV['PATHEXT'] as it is extremely unlikely that PATHEXT will contain non-us english charaters i.e.  *.EXE or *.CMD

Iristyle commented on lib/puppet/util.rb:
```diff
@@ -25,6 +25,67 @@ module Util
+    end
+  end
+  # TODO use alias_method instead
+  def self.env(name,mode = nil)
```
I realize you asked me about this the other day, but the ask seems clearer in the code to me now.  I think that we should use a hash-like structure, similar to `ENV` to allow getting / setting through the same interface *or* we use a `get` / `set` dual.

glennsarti replied:
So the big question is, do we 'patch' ENV or implement a new module called Puppet::Env or Puppet::Util::Env?

We can't use Puppet::Environments as that's for puppet environment

Iristyle replied:
While there are some appealing reasons to patch `ENV`, I'd definitely prefer to find another way.  I think the simplest thing that works here is to just change `self.env` to `self.get_env`, and add a note about the Ruby bug / that the code can be removed when we're on 2.3.

Iristyle commented on lib/puppet/util.rb:
```diff
@@ -25,6 +25,67 @@ module Util
+  end
+
+  def self.clear_environment
+    ENV.clear
```
Does this work OK on Windows?  Is it ever called on Windows?

glennsarti replied:
Yes it works on windows.  It is called in the spec_helper and in fact that was the only reason I implemented platform specific implementation

glennsarti commented:
Unfortunately the terminating sequence can be either a single or double null widechar (2 or 4 byte).  I suppose a simple function param could be used to get either a 2 or 4 byte integer (uint32)

glennsarti commented:
I've updated the memory usage commit with suggestions from @Iristyle and @peterhuene. Memory usage is now much smaller.  Also added tests to ensure against regressions.

ferventcoder commented on lib/puppet/util.rb:
```diff
@@ -25,26 +25,49 @@ module Util
-  # Rubys ENV tries to roundtrip them through the local codepage, which can
-  # cause encoding problems - underlying helpers use Windows APIs on Windows
-  # see https://bugs.ruby-lang.org/issues/8822
-  def self.withenv(hash, mode = :posix)
```
Any reason this is not deprecated? I don't see it behind a private switch so IMHO it must be deprecated and removed on a major version bump.

ferventcoder replied:
Nevermind, found it still here.

Iristyle commented:
FYI - I've been trying to run the full spec suite against 3 different machines (fresh Windows 10 devbox with Ruby 2.1.8, Windows 8.1 dev box with 2.1.8, 2008R2 dev box with 2.1.5)... memory utilization goes upward fast, and they all segfault pretty quickly.  Going to hit the bed, and hopefully something will seep into my head overnight ;0

### Windows 8.1

```
ruby 2.1.8p440 (2015-12-16 revision 53160) [x64-mingw32]

-- Control frame information -----------------------------------------------
c:0053 p:---- s:0233 e:000232 CFUNC  :initialize
c:0052 p:---- s:0231 e:000230 CFUNC  :new
c:0051 p:0076 s:0226 e:000225 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/windows/error.rb:36
c:0050 p:0031 s:0219 e:000215 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/windows/error.rb:14 [FINISH]
c:0049 p:---- s:0210 e:000209 CFUNC  :new
c:0048 p:0065 s:0206 e:000204 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/windows/file.rb:97
c:0047 p:0011 s:0200 e:000199 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/windows/file.rb:183
c:0046 p:0085 s:0195 e:000194 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/file_system/windows.rb:15
c:0045 p:0014 s:0191 e:000190 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/file_system.rb:158

c:0044 p:0028 s:0187 e:000186 BLOCK  C:/Users/Parity/Documents/source/puppet/lib/puppet/util/autoload.rb:86 [FINISH]
c:0043 p:---- s:0184 e:000183 IFUNC
c:0042 p:---- s:0182 e:000181 CFUNC  :each
c:0041 p:---- s:0180 e:000179 CFUNC  :find
c:0040 p:0029 s:0177 e:000176 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/autoload.rb:86
c:0039 p:0013 s:0171 e:000170 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/autoload.rb:55
c:0038 p:0031 s:0163 e:000162 BLOCK  C:/Users/Parity/Documents/source/puppet/lib/puppet/util/autoload.rb:74 [FINISH]
c:0037 p:---- s:0159 e:000158 CFUNC  :each
c:0036 p:0016 s:0156 e:000155 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/autoload.rb:72
c:0035 p:0017 s:0151 e:000150 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/util/autoload.rb:203
c:0034 p:0009 s:0147 e:000146 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface.rb:174
c:0033 p:0066 s:0144 e:000143 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface.rb:65
c:0032 p:0027 s:0137 E:001ed8 TOP    C:/Users/Parity/Documents/source/puppet/lib/puppet/face/catalog.rb:3 [FINISH]
c:0031 p:---- s:0135 e:000134 CFUNC  :require
c:0030 p:0049 s:0131 e:000130 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface/face_collection.rb:100
c:0029 p:0009 s:0124 e:000123 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface/face_collection.rb:56
c:0028 p:0031 s:0118 e:000117 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface/face_collection.rb:17
c:0027 p:0011 s:0113 e:000112 BLOCK  C:/Users/Parity/Documents/source/puppet/lib/puppet/interface/face_collection.rb:10 [FINISH]
c:0026 p:---- s:0110 e:000109 CFUNC  :each
c:0025 p:0035 s:0107 e:000106 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface/face_collection.rb:10
c:0024 p:0017 s:0103 e:000102 METHOD C:/Users/Parity/Documents/source/puppet/lib/puppet/interface.rb:37
c:0023 p:0015 s:0100 e:000099 BLOCK  C:/Users/Parity/Documents/source/puppet/spec/integration/faces/documentation_spec.rb:19 [FINISH]
c:0022 p:---- s:0098 e:000097 CFUNC  :module_exec
c:0021 p:0042 s:0095 e:000094 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:385
c:0020 p:0143 s:0087 e:000086 LAMBDA C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:255 [FINISH]
c:0019 p:0017 s:0079 E:000b18 BLOCK  C:/Users/Parity/Documents/source/puppet/spec/integration/faces/documentation_spec.rb:18 [FINISH]
c:0018 p:---- s:0077 e:000076 CFUNC  :module_exec
c:0017 p:0042 s:0074 e:000073 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:385
c:0016 p:0143 s:0066 e:000065 LAMBDA C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:255 [FINISH]
c:0015 p:0032 s:0058 e:000056 LAMBDA C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/dsl.rb:43 [FINISH]
c:0014 p:0023 s:0053 e:000052 LAMBDA C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/dsl.rb:82 [FINISH]
c:0013 p:0023 s:0049 E:001438 TOP    C:/Users/Parity/Documents/source/puppet/spec/integration/faces/documentation_spec.rb:5 [FINISH]
c:0012 p:---- s:0047 e:000046 CFUNC  :load
c:0011 p:0024 s:0043 e:000042 BLOCK  C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/configuration.rb:1361 [FINISH]
c:0010 p:---- s:0039 e:000038 CFUNC  :each
c:0009 p:0027 s:0036 e:000035 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/configuration.rb:1359
c:0008 p:0051 s:0033 e:000032 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:106
c:0007 p:0011 s:0028 e:000027 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:92
c:0006 p:0097 s:0023 e:000022 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:78
c:0005 p:0024 s:0016 e:000015 METHOD C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:45
c:0004 p:0023 s:0012 e:000011 TOP    C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/exe/rspec:4 [FINISH]
c:0003 p:---- s:0010 e:000009 CFUNC  :load
c:0002 p:0135 s:0006 E:000948 EVAL   C:/Users/Parity/Documents/source/puppet/.bundle/gems/ruby/2.1.0/bin/rspec:23 [FINISH]
c:0001 p:0000 s:0002 E:000d48 TOP    [FINISH]
```


### Windows 10

```
ruby 2.1.8p440 (2015-12-16 revision 53160) [x64-mingw32]

-- Control frame information -----------------------------------------------
c:0053 p:---- s:0233 e:000232 CFUNC  :initialize
c:0052 p:---- s:0231 e:000230 CFUNC  :new
c:0051 p:0076 s:0226 e:000225 METHOD C:/source/puppet/lib/puppet/util/windows/error.rb:36
c:0050 p:0031 s:0219 e:000215 METHOD C:/source/puppet/lib/puppet/util/windows/error.rb:14 [FINISH]
c:0049 p:---- s:0210 e:000209 CFUNC  :new
c:0048 p:0065 s:0206 e:000204 METHOD C:/source/puppet/lib/puppet/util/windows/file.rb:97
c:0047 p:0011 s:0200 e:000199 METHOD C:/source/puppet/lib/puppet/util/windows/file.rb:183
c:0046 p:0085 s:0195 e:000194 METHOD C:/source/puppet/lib/puppet/file_system/windows.rb:15
c:0045 p:0014 s:0191 e:000190 METHOD C:/source/puppet/lib/puppet/file_system.rb:158
c:0044 p:0028 s:0187 e:000186 BLOCK  C:/source/puppet/lib/puppet/util/autoload.rb:95 [FINISH]
c:0043 p:---- s:0184 e:000183 IFUNC
c:0042 p:---- s:0182 e:000181 CFUNC  :each
c:0041 p:---- s:0180 e:000179 CFUNC  :find
c:0040 p:0029 s:0177 e:000176 METHOD C:/source/puppet/lib/puppet/util/autoload.rb:95
c:0039 p:0013 s:0171 e:000170 METHOD C:/source/puppet/lib/puppet/util/autoload.rb:64
c:0038 p:0031 s:0163 e:000162 BLOCK  C:/source/puppet/lib/puppet/util/autoload.rb:83 [FINISH]
c:0037 p:---- s:0159 e:000158 CFUNC  :each
c:0036 p:0016 s:0156 e:000155 METHOD C:/source/puppet/lib/puppet/util/autoload.rb:81
c:0035 p:0017 s:0151 e:000150 METHOD C:/source/puppet/lib/puppet/util/autoload.rb:208
c:0034 p:0009 s:0147 e:000146 METHOD C:/source/puppet/lib/puppet/interface.rb:174
c:0033 p:0066 s:0144 e:000143 METHOD C:/source/puppet/lib/puppet/interface.rb:65
c:0032 p:0027 s:0137 E:000e28 TOP    C:/source/puppet/lib/puppet/face/catalog.rb:3 [FINISH]
c:0031 p:---- s:0135 e:000134 CFUNC  :require
c:0030 p:0049 s:0131 e:000130 METHOD C:/source/puppet/lib/puppet/interface/face_collection.rb:100
c:0029 p:0009 s:0124 e:000123 METHOD C:/source/puppet/lib/puppet/interface/face_collection.rb:56
c:0028 p:0031 s:0118 e:000117 METHOD C:/source/puppet/lib/puppet/interface/face_collection.rb:17
c:0027 p:0011 s:0113 e:000112 BLOCK  C:/source/puppet/lib/puppet/interface/face_collection.rb:10 [FINISH]
c:0026 p:---- s:0110 e:000109 CFUNC  :each
c:0025 p:0035 s:0107 e:000106 METHOD C:/source/puppet/lib/puppet/interface/face_collection.rb:10
c:0024 p:0017 s:0103 e:000102 METHOD C:/source/puppet/lib/puppet/interface.rb:37
c:0023 p:0015 s:0100 e:000099 BLOCK  C:/source/puppet/spec/integration/faces/documentation_spec.rb:19 [FINISH]
c:0022 p:---- s:0098 e:000097 CFUNC  :module_exec
c:0021 p:0042 s:0095 e:000094 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:385
c:0020 p:0143 s:0087 e:000086 LAMBDA C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:255 [FINISH]
c:0019 p:0017 s:0079 E:0018b8 BLOCK  C:/source/puppet/spec/integration/faces/documentation_spec.rb:18 [FINISH]
c:0018 p:---- s:0077 e:000076 CFUNC  :module_exec
c:0017 p:0042 s:0074 e:000073 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:385
c:0016 p:0143 s:0066 e:000065 LAMBDA C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/example_group.rb:255 [FINISH]
c:0015 p:0032 s:0058 e:000056 LAMBDA C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/dsl.rb:43 [FINISH]
c:0014 p:0023 s:0053 e:000052 LAMBDA C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/dsl.rb:82 [FINISH]
c:0013 p:0023 s:0049 E:000d48 TOP    C:/source/puppet/spec/integration/faces/documentation_spec.rb:5 [FINISH]
c:0012 p:---- s:0047 e:000046 CFUNC  :load
c:0011 p:0024 s:0043 e:000042 BLOCK  C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/configuration.rb:1361 [FINISH]
c:0010 p:---- s:0039 e:000038 CFUNC  :each
c:0009 p:0027 s:0036 e:000035 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/configuration.rb:1359
c:0008 p:0051 s:0033 e:000032 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:106
c:0007 p:0011 s:0028 e:000027 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:92
c:0006 p:0097 s:0023 e:000022 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:78
c:0005 p:0024 s:0016 e:000015 METHOD C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/lib/rspec/core/runner.rb:45
c:0004 p:0023 s:0012 e:000011 TOP    C:/source/puppet/.bundle/gems/ruby/2.1.0/gems/rspec-core-3.4.3/exe/rspec:4 [FINISH]
c:0003 p:---- s:0010 e:000009 CFUNC  :load
c:0002 p:0135 s:0006 E:0018e8 EVAL   C:/source/puppet/.bundle/gems/ruby/2.1.0/bin/rspec:23 [FINISH]
c:0001 p:0000 s:0002 E:0004e8 TOP    [FINISH]
```

Iristyle commented:
FWIW - its the same failure spot each time - https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/error.rb#L36

Iristyle commented:
Of course, I still had `page heap` validation set with `gflags` in all instances :scream_cat:

Iristyle commented:
Closed in favor of #4762
